use exponential backoff retry policy

d2r opened the pull request:
Use com.netflix.curator.retry.ExponentialBackoffRetry instead of com.netflix.curator.retry.RetryNTimes for ZK connections.

Commit:
use exponential backoff retry policy

nathanmarz commented:
Did you run into something in particular that required this?

d2r commented:
Reiterating Andy's rationale for the change:

With a large cluster, after Zookeeper is not accessible for a period of time, we found Zookeeper can get overwhelmed when trying to recover.  The exponential back-off helps in those situations.

nathanmarz commented on src/jvm/backtype/storm/utils/Utils.java:
```diff
@@ -299,7 +299,7 @@ public static CuratorFramework newCurator(Map conf, List<String> servers, Object
                     .connectionTimeoutMs(Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_CONNECTION_TIMEOUT)))
                     .sessionTimeoutMs(Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_SESSION_TIMEOUT)))
-                    .retryPolicy(new RetryNTimes(Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_RETRY_TIMES)), Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_RETRY_INTERVAL))));
+                    .retryPolicy(new ExponentialBackoffRetry(Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_RETRY_TIMES)), Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_RETRY_INTERVAL))));
```
According to http://netflix.github.com/curator/doc/com/netflix/curator/retry/ExponentialBackoffRetry.html, these arguments are reversed here.

nathanmarz commented:
This is a good idea in principle. Some questions:
1. What do the time intervals look like exactly? If your base is 100ms, what will the next 10 tries look like? 200ms, 400ms, 800ms, 1.6s, 3.2s, etc?
2. Is there a way to set a ceiling on the retry time? Is this even desirable? Seems like some serious problems could arise if a worker is waiting 30s to retry an operation – like for example, a worker timeout.

d2r commented:
1. They would look like: 100~200ms, 100~400ms, 100~800ms, etc.  With a base of 100ms, the interval for a given `retryCount` is random between 100ms ~ 2^(`retryCount` + 1).  So it is randomized, with the upper limit increasing exponentially.
2. Yes we can, and I agree that probably would be a good idea. It defaults to `Integer.MAX_VALUE`.  What would be a good ceiling?

d2r commented:
Further notes on the ceiling.  The version of curator we pull in now (1.0.1) does not have a ceiling.  We'd need at least 1.2.6.

Also, in the curator git revision comments there is a note that they changed the ceiling from 60000ms to `Integer.MAX_VALUE` to avoid "breaking existing code."

d2r commented:
What we really want is BoundedExponentialBackoffRetry, which was added in curator 1.2.5.

But I note that there is a significant hurdle to pulling in a newer curator: #225, especially it's dependency on zookeeper 3.4.5 #467

Commit:
Use a ceiling on ExponentialBackoffRetry

Trying to avoid pulling in a newer netflix curator

Adds lein-junit and junit dependencies with test.

Adds new configurable maximum ceiling for the retry interval, defaults
to 30s.

d2r commented:
Put a configurable ceiling on the backoff retry, defaults to 30s, but we can discuss what it should be.
Added JUnit and a test for the change.

Trying to avoid pulling in a new Curator.  We may not want to go down this road either...

nathanmarz commented:
Yea, we've had some trouble with the new ZK client library that we're trying to hash out. 

As for the diff – let's keep the tests in Clojure. Just add a new util_test.clj file into test/clj/backtype/storm/

d2r commented:
Sure, I can rewrite the test. 

Curious, what is the barrier to using JUnit tests?

nathanmarz commented:
JUnit adds no additional functionality compared to what we can already do with clojure.test, so it's simpler to stick with the existing unit testing framework. This also reduces dependencies of the project and makes it simpler to run the unit tests (just lein test).

Commit:
Change JUnit test to clojure.test test.

Pulled out the retry policy into its own class.

d2r commented:
OK, rewritten with a clojure test.

nathanmarz commented on test/clj/backtype/storm/util_test.clj:
```diff
@@ -0,0 +1,33 @@
+    (is (= (.getBaseSleepTimeMs retry) expected_interval)) 
+    (is (= (.getN retry) expected_retries)) 
+    (is (= (.getMaxRetryInterval retry) expected_ceiling)) 
+    ; It would be very unlikely for this to fail three times.
```
What does this mean exactly? Isn't the function stateless, so can't you just test the value once?

d2r replied:
The result is psuedo-randomly distributed, so there is a slim chance with these values that working code could trip the assertion.

I could trivially set the ceiling to be less than the base, and that would guarantee the assertion is true.

nathanmarz commented:
OK – I think we're close to finishing this. First off, let's get rid of the project.clj changes as those are no longer necessary. I also made one more comment in the code,

Commit:
Revert changes for JUnit dependency

d2r commented:
Reverted the project.clj change, sorry about that.  Replied to your in-line comment.

Commit:
Change ceiling test to be deterministic

Commit:
Rename util -> utils in new test

Commit:
Remove trailing white-space

nathanmarz commented:
Merged. I changed the config to explicitly say millis.
